Issue #3483 Add cert retrieval for requests

dusktreader opened the pull request:
The authenticator.py code already retrieves credentials from the config
for every request based on url matching. This change makes the
authenticator also retrieve certs from the config for each request based
on url matching.

Also includes unit tests.

# Pull Request Check List

Resolves: #3483 

<!-- This is just a reminder about the most common mistakes. Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](https://python-poetry.org/docs/contributing/) at least once, it will save you unnecessary review cycles! -->

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code.

<!-- If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing! -->

tbennett6421 commented:
Forgive my ignorance. Would this work for `poetry publish`?

dusktreader commented:
@tbennett6421 Yes, that's specifically what this is for.

GooseYArd commented:
I just noticed that when cert_verify is called by _download_archive, cert's type is Path(), so the test

if not isinstance(cert, basestring): 

yields False, and we mistakenly treat the cert as a list.

Oddly, when cert_verify winds up being called by legacy_repository.find_links_for_package, the certs name is passed as a string.

I'll see if I can figure out a fix

GooseYArd commented:
This seems to do the trick:

diff --git a/poetry/installation/authenticator.py b/poetry/installation/authenticator.py
index 8b57c9c..8fd699f 100644
--- a/poetry/installation/authenticator.py
+++ b/poetry/installation/authenticator.py
@@ -72,8 +72,8 @@ class Authenticator(object):
         stream = kwargs.get("stream")
 
         certs = self.get_certs_for_url(url)
-        verify = kwargs.get("verify", certs.get("verify"))
-        cert = kwargs.get("cert", certs.get("cert"))
+        verify = str(kwargs.get("verify", certs.get("verify")))
+        cert = str(kwargs.get("cert", certs.get("cert")))
 
         settings = session.merge_environment_settings(
             prepared_request.url, proxies, stream, verify, cert

the requests module will not accept Path instances for either cert or verify, but it will accept cert as a tuple of (cert, key), in cases where the cert and key are not combined into a single file.  I'm assuming that poetry expects the combined format since pip has the same expectation. I'm guessing that if poetry ever opts to support separate cert/key files, it would entail a different key in auth.yaml, so I doubt we'll need to check here whether cert is a tuple or not.

dusktreader commented:
@GooseYArd it seems like you could just apply `str()` to any value of cert and verify, so maybe the if statements aren't necessary?

GooseYArd commented:
> @GooseYArd it seems like you could just apply `str()` to any value of cert and verify, so maybe the if statements aren't necessary?

hah funny you mention it- that was what I did in the initial patch without realizing that the values will be None if no cert/key or calist are specified, and request's cert verification was trying to locate the literal 'None' :)

dusktreader commented:
@GooseYArd good point there. i'll apply this patch

BrutalSimplicity commented:
Any chance this will be merged anytime soon? I'd love to promote Poetry as the default package manager, but work with enterprises that have an internal PyPi repository, and am unable to pull those packages without hacky workarounds.

obviousrebel commented:
The issue resolved here is also blocking adoption for our project

GooseYArd commented:
@abn this was the second of the two private repo + cert related issues that were preventing us from using Artifactory pypi servers. I'm hoping it can be reviewed for 1.2, but am not sure who to ping. We've been using this PR internally for the past few months without issue, fwiw.

abn commented on poetry/installation/authenticator.py:
```diff
@@ -145,6 +159,29 @@ def _get_credentials_for_netloc_from_config(
+            self._get_certs_for_netloc_from_config(netloc),
+        )
+
+    def _get_repository_netlocs(self):  # type: () -> Generator[[str, str], None, None]
```
We can drop compatibily type hints now.

abn reviewed (changes requested):
Over all looks okay; can we get this rebased and compatibility code removed please?

abn commented:
@GooseYArd @dusktreader apologies this slipped through the radar. If you can get it rebased and updated, I can get it reviewed and merged.

GooseYArd commented:
@abn thanks a bunch, if @dusktreader doesn't beat me to it I'll make the updates you requested tomorrow AM

dusktreader commented:
@abn Rebased and force-pushed! thanks

@GooseYArd got it covered, I think

surry commented on poetry/installation/authenticator.py:
```diff
@@ -61,8 +67,16 @@ def request(self, method: str, url: str, **kwargs: Any) -> requests.Response:
+        verify = kwargs.get("verify", certs.get("verify"))
+        cert = kwargs.get("cert", certs.get("cert"))
+
+        if isinstance(cert, pathlib.PurePath):
```
I changed this check to `if cert:` and things seem to work as expected for me on Ubuntu with only a client cert specified (verify should be `None` for me).

surry commented:
@dusktreader thanks for this PR, also a blocker for us since we use a private repository with client certificates. I manually made these changes to my Poetry 1.1.6 install and my private repo is working now. Are you able to update and merge this soon? Would love to get this fix into a version I can update to with `poetry self update`.

dusktreader commented:
@abn Applied suggested changes and pushed commit. Thanks!

surry commented on poetry/installation/authenticator.py:
```diff
@@ -61,8 +67,16 @@ def request(self, method: str, url: str, **kwargs: Any) -> requests.Response:
+        if cert is not None:
+            cert = str(cert)
+
+        if isinstance(verify, pathlib.PurePath):
```
```suggestion
        if verify is not None:
```

Same as the change @abn suggested for the block above.

surry commented on poetry/installation/authenticator.py:
```diff
@@ -138,6 +152,29 @@ def _get_credentials_for_netloc_from_config(
+
+        return credentials
+
+    def get_certs_for_url(self, url: str) -> Dict[str, pathlib.PosixPath]:
```
```suggestion
    def get_certs_for_url(self, url: str) -> Dict[str, pathlib.Path]:
```

Think the same change was made for `_get_certs_for_netloc_from_config()`

dusktreader replied:
Applied your suggestions. Thank you!

surry reviewed (commented):
Minor suggestions along the lines of @abn's ones. Also, I quickly looked at why the Python 3.10 tests failed and it might be due to this:

https://github.com/pytest-dev/pytest/issues/8539

Not sure what the fix is here, if any...

@abn would it be possible to merge this in for the upcoming 1.2.0a1 release?

abn commented on poetry/installation/authenticator.py:
```diff
@@ -138,6 +152,29 @@ def _get_credentials_for_netloc_from_config(
+        return credentials
+
+    def get_certs_for_url(self, url: str) -> Dict[str, pathlib.Path]:
+        parsed_url = urllib.parse.urlsplit(url)
+
+        netloc = parsed_url.netloc
+
+        return self._certs.setdefault(
+            netloc,
+            self._get_certs_for_netloc_from_config(netloc),
+        )
```
```suggestion
        return self._certs.setdefault(
            urllib.parse.urlsplit(url).netloc,
            self._get_certs_for_netloc_from_config(netloc),
        )
```

abn replied:
Ah yes, missed that I would just keep the netloc var then.

abn commented on poetry/installation/authenticator.py:
```diff
@@ -61,8 +67,16 @@ def request(self, method: str, url: str, **kwargs: Any) -> requests.Response:
-        cert = kwargs.get("cert")
+
+        certs = self.get_certs_for_url(url)
+        verify = kwargs.get("verify", certs.get("verify"))
+        cert = kwargs.get("cert", certs.get("cert"))
```
```suggestion
        verify = kwargs.get("verify") or certs.get("verify")
        cert = kwargs.get("cert") or certs.get("cert")
```

This should handle potential cases where `None` is explicitly passed through.

dusktreader replied:
Addressed in commit 26b4ce7731dea5c6ce7fb1980dae4a2b52e66ce3

abn commented on tests/installation/test_authenticator.py:
```diff
@@ -201,3 +202,50 @@ def test_authenticator_uses_env_provided_credentials(
+            "repositories": {"foo": {"url": "https://foo.bar/simple/"}},
+            "http-basic": {"foo": {"username": "bar", "password": "baz"}},
+            "certificates": {
+                "foo": {"cert": "/path/to/cert", "client-cert": "/path/to/client-cert"}
+            },
+        }
+    )
+
+    authenticator = Authenticator(config, NullIO())
+    session_send = mocker.patch.object(authenticator.session, "send")
+    authenticator.request("get", "https://foo.bar/files/foo-0.1.0.tar.gz")
+    kwargs = session_send.call_args[1]
+    assert pathlib.Path(kwargs["verify"]) == pathlib.Path("/path/to/cert")
+    assert pathlib.Path(kwargs["cert"]) == pathlib.Path("/path/to/client-cert")
+
+
+def test_authenticator_uses_provided_certs_instead_of_config_certs(
+    config, mock_remote, http, mocker
+):
+    config.merge(
+        {
+            "repositories": {"foo": {"url": "https://foo.bar/simple/"}},
+            "http-basic": {"foo": {"username": "bar", "password": "baz"}},
+            "certificates": {
+                "foo": {"cert": "/path/to/cert", "client-cert": "/path/to/client-cert"}
+            },
+        }
+    )
+
+    authenticator = Authenticator(config, NullIO())
+    session_send = mocker.patch.object(authenticator.session, "send")
+    authenticator.request(
+        "get",
+        "https://foo.bar/files/foo-0.1.0.tar.gz",
+        verify="/path/to/provided/cert",
+        cert="/path/to/provided/client-cert",
+    )
+    kwargs = session_send.call_args[1]
+    assert pathlib.Path(kwargs["verify"]) == pathlib.Path("/path/to/provided/cert")
+    assert pathlib.Path(kwargs["cert"]) == pathlib.Path("/path/to/provided/client-cert")
```
```suggestion
@pytest.mark.parametrize("cert,client_cert", [
    (None, None),
    ("/path/to/provided/cert", "/path/to/provided/client-cert")
])
def test_authenticator_uses_certs_from_config_if_not_provided(
    cert, client_cert, config, mock_remote, http, mocker
):
    configured_cert = "/path/to/cert"
    configured_client_cert = "/path/to/client-cert"
    config.merge(
        {
            "repositories": {"foo": {"url": "https://foo.bar/simple/"}},
            "http-basic": {"foo": {"username": "bar", "password": "baz"}},
            "certificates": {
                "foo": {"cert": configured_cert, "client-cert": configured_client_cert}
            },
        }
    )

    authenticator = Authenticator(config, NullIO())
    session_send = mocker.patch.object(authenticator.session, "send")
    authenticator.request("get", "https://foo.bar/files/foo-0.1.0.tar.gz", verify=cert, cert=client_cert)
    kwargs = session_send.call_args[1]
    
    assert Path(kwargs["verify"]) == Path(cert or configured_cert)
    assert Path(kwargs["cert"]) == Path(client_cert or configured_client_cert)
```
Nothing wroing with the implementation, just wanted to avoid some duplicated code.

dusktreader replied:
I think it makes sense to have the tests separated since the test name clearly defines the conditions under test. In this suggestion, the parameterization makes the test cases less clear.

abn replied:
I am not wedded to parametrised tests. But I am much more keen on adding to the already large sets of duplicated code in our test suites. It is becoming harder to maintain. Additionally you can add a comment next to the parameters if further clarity is required. Personally, I do think the parameters are clear enough.

dusktreader replied:
OK. Will update

dusktreader replied:
This has been updated as requested

abn commented on tests/installation/test_authenticator.py:
```diff
@@ -1,3 +1,4 @@
+import pathlib
```
```suggestion
from pathlib import Path
```

dusktreader replied:
There's a few reasons to prefer `import pathlib` to `from pathlib import Path`. The primary one is that it follows the import pattern used in this module already with requests, pytest, and other. Secondly, `pathlib.Path` is more clear and explicit the type of `Path` object being used since the name Path is common and can be overloaded. The last reason, though it isn't really relevant here, is that using the `import path` style makes it much easier to mock internal functions in pathlib if necessary.

abn replied:
I would say those reasons are rather subjective and not necessarily universal. Further, I do not think we use anything other than `pathlib.Path` as `Path` through the code base. I recommended the change as this is how `Path` is import through the rest of the code. As for mocking, I suspect we will have to patch anyway within `poetry.*` in most cases.

dusktreader replied:
OK. Will update

GooseYArd replied:
Since 1.2.0 looks imminent I thought I'd bump this thread- this is the last cert-related issue that is blocking my company from being able to use poetry generally for our internal packaging, and since I've already converted a number of our packages (and have been using a private fork with a couple of cert fixes) I'm desperate to get this change merged. Is there a compromise we could reach here?

dusktreader replied:
I will try to get some cycles to work on this today. The latest requests for changes broke a bunch of unit tests, so I will have to figure out what happened there. Sorry it's been slow to merge.

dusktreader replied:
Updates pushed. Tests passed locally; will see what the pipeline does when it's greenlit to run

surry replied:
The python3.6 tests on FreeBSD failed with a strange (or missing) error: https://github.com/python-poetry/poetry/pull/3490/checks?check_run_id=2657871763

Maybe that test just needs to be re-run and the other workflows could be approved by a maintainer see if there are other issues this PR needs to address? @abn

dusktreader replied:
Yes, that failure seems to be unrelated to my change. I will try to act on any more feedback on this quickly, since folks are waiting on this feature.

dusktreader replied:
This has been updated as requested.

abn commented on poetry/installation/authenticator.py:
```diff
@@ -1,9 +1,12 @@
 import logging
+import pathlib
```
```suggestion
from pathlib import Path
```

dusktreader replied:
This has been adjusted as requested

abn reviewed (changes requested):
Functionally looks okay. Some minor suggestions added.

dusktreader commented on poetry/installation/authenticator.py:
```diff
@@ -138,6 +152,29 @@ def _get_credentials_for_netloc_from_config(
+                credentials = (auth["username"], auth["password"])
+                break
+
+        return credentials
```
Addressed in commit 26b4ce7

dusktreader commented:
@abn Can you trigger the pipelines, please? Also, it seems like the "Tests / FreeBSD / PYTHON:python3.6" is failing due to an unrelated issue. Also, I've addressed your comments above

dusktreader commented:
Can we get a maintainer to look at this please? Just waiting for pipelines

surry commented:
I pinged some of the devs on Discord to get the pipelines going, so you might be able to get more traction there. 

@abn would you be able to take another look now that the checks pass? This PR would really help users that rely on client certificates to access private Pypi repositories.

obviousrebel commented:
Any idea when this would be implemented? Project stuck at 1.0.10 until this is fixed for use of our private repo.

jakeri commented:
Any updates on this one?

We wanted to use poetry for our internal build but we need the client cert functional. No we have to do a really convoluted build step with extracting requirements.txt, etc.

dusktreader commented:
Rebased to fix merge conflict.

dusktreader commented:
There was a refactor of the same module that this PR worked in and the rebase broke some things. Fixed that, and we should be good to go now.

dusktreader commented on poetry/utils/authenticator.py:
```diff
@@ -168,7 +181,7 @@ def _get_credentials_for_netloc(
         credentials = (None, None)
 
-        for repository_name in self._config.get("repositories", []):
+        for (repository_name, repository_netloc) in self._get_repository_netlocs():
```
Addressed in latest commit.

dusktreader commented on tests/utils/test_authenticator.py:
```diff
@@ -249,3 +251,39 @@ def test_authenticator_uses_env_provided_credentials(
+    "cert,client_cert",
+    [
+        (None, None),
+        ("/path/to/provided/cert", "path/to/provided/client-cert"),
```
Addressed in latest commit.

LarsStegman commented:
Any updates on this?

dusktreader commented:
> Any updates on this?

Still awaiting approval on the last round of update/rebase. Not much more I can do.

LarsStegman commented:
@abn We are starting to run into some problems with compatibility between the version of Poetry we're stuck at and Python 3.10. Do you have time to look at this PR?

dusktreader commented:
Rebased and force pushed again.

@abn or anyone else: does this have any traction?

LarsStegman commented:
@abn @finswimmer @neersighted @sdispater @stephsamson  Could one of the maintainers please look at this? I'm running into problems with multiple of our projects due to being stuck at an old version of Poetry.

BrutalSimplicity commented:
The completion of this PR is holding up full adoption for 100s of developers at my company. It's been in-progress for over a year. Any chance we can see this completed soon?

Is there some method to the prioritization of these requests? It seems like being able to use a certificate for requests made by Poetry would be a major blocker to widespread adoption for many professional developers.

ITProKyle commented:
Would it be possible to release this as a patch to 1.1 to get it out the door faster? It is a major blocker for many and is desperately needed ASAP.

dusktreader commented:
I simply don't have time to do this right now, and it's not in my current critical path. If someone else has time to pick up the mantle and rebase it, I (and apparently many others) would greatly appreciate it.

chriswhite199 commented:
For those blocked by this, can you not just set CURL_CA_BUNDLE env variable before running poetry to your CA bundle file?

This is what I'm doing for hosted Artifactory pypi mirror and private repos.

ITProKyle commented:
@chriswhite199 - yes and no.

Yes, that workaround works for **me** as someone that is familiar with poetry and its quirks.

No, that workaround will not work for the >100 developers I work with that have never used poetry.

chriswhite199 commented:
@ITProKyle - I hear you, and we just add a poetry wrapper script in the base of our repositories to do this - It handles detection of whether poetry is installed globally (and if not pip installs it into a .poetry-venv folder). Developers are taught to use it (in the same way they might use a [maven](https://maven.apache.org/wrapper/) or [gradle](https://docs.gradle.org/current/userguide/gradle_wrapper.html) wrapper if they were coding in Java).

papb commented:
> For those blocked by this, can you not just set CURL_CA_BUNDLE env variable before running poetry to your CA bundle file?
> 
> This is what I'm doing for hosted Artifactory pypi mirror and private repos.

Oohhhhhhh thank you for teaching me this. Where did you find about this `CURL_CA_BUNDLE` env variable? Where is this documented? @chriswhite199
